### ISSUE TITLE
Exhaustiveness check with GADT doesn't work

### ISSUE THREAD

**Issue by shohei909** [HaxeFoundation/haxe#4689]
```haxe
class Main 
{
	static function main() 
	{
		trace(test(Gadt.B)); // NaN
	}
	
	private static function test<T>(a:Gadt<T>):T
	{
		return switch (a) 
		{
			case A:
				"test";
		}
	}
}

enum Gadt<T> {
	A:Gadt<String>;
	B:Gadt<Float>;
}
```

In Haxe 3.1.3, this cause "Unmatched patterns: B" error, but in Haxe 3.2.0 or later (also development build), compilation succeed.

**Comment by Simn** [comment]
Bisected to https://github.com/HaxeFoundation/haxe/commit/b3eab6f78312ddb045b5e68d2e1d927ab5bfe8e0

**Comment by Simn** [comment]
My fix for #2778 wasn't good: We do indeed want to use the subject type, but we want to keep it constant per-case. Once we enter a new case we have to allow type parameters to be bound again, otherwise we miss constructors like we do here.

That's actually not so easy to implement properly with the current matcher architecture. I'll have to think about how to address this.
